PyInstaller/loader/pyimod04_ctypes.py

```diff
@@ -0,0 +1,105 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2005-2021, PyInstaller Development Team.
+#
+# Distributed under the terms of the GNU General Public License with exception
+# for distributing bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+
+"""
+Hooks to make ctypes.CDLL, .PyDLL, etc. look in sys._MEIPASS first.
+"""
+
+import sys
+
+
+def install():
+    """Install the hooks.
+
+    This must be done from a function as opposed to at module-level,
+    because when the module is imported/executed, the import machinery
+    is not completely set up yet.
+    """
+
+    import os
+
+    try:
+        import ctypes
+    except ImportError:
+        # ctypes is not included in the frozen application
+        return
+
+    def _frozen_name(name):
+        if name:
+            frozen_name = os.path.join(sys._MEIPASS, os.path.basename(name))
+            if os.path.exists(frozen_name) and not os.path.isdir(frozen_name):
+                name = frozen_name
+        return name
```

Comment:
Can we change this so that original full paths have top priority? 
```suggestion
        if name and not os.path.exists(name):
            frozen_name = os.path.join(sys._MEIPASS, os.path.basename(name))
            if os.path.exists(frozen_name) and not os.path.isdir(frozen_name):
                name = frozen_name
        return name
```

I vaguely remember a case where two different versions of the same DLL got bundled and this snippet of code got in the way by forcing the wrong DLL in the root of the application to be used.

Reply:
That's a good idea. The question is, though, do we do it in this PR, or a separate one (with its own news entry)? Even if it were done here, I'd prefer to keep it in a separate commit...

Reply:
Hmm, different commit definitely. Don't see any harm in using this PR. I'll see if I can find the issue where the problem happened and you can use its number in the news entry rather than this semi-relevant PR.

Reply:
Edit: I have absolutely no idea where that issue has wandered off to. In which case it probably should be a separate PR.

Reply:
You weren't thinking of #5583 or #3993, were you? Without diverting extensions to `lib-dynload`, those two would still occur even if priorities were reversed (and would occur as long as we also try to check `sys._MEIPASS`).

Reply:
Ah yes it was #5583.

Reply:
Although I remember something else about conda and zlib...

Reply:
Ah, #3659. But it's not relevant...

Reply:
> Ah, #3659. But it's not relevant...

Indeed, doesn't seem related to `ctypes`.